024-revise-operations.md

```diff
@@ -0,0 +1,261 @@
+Standardize Operation Formatting
+================================
+
+Background
+----------
+
+This proposal is editorial in nature: it does not intend to change the
+meaning of any DWARF constructs, but merely to standardize the
+presentation of the existing DWARF operations to facilitate
+understanding.
+
+It is believed that tool authors referring to the the standard would
+find it much easier to understand if each operand followed a standard
+template, rather than explaining these in the textual description of
+the operand. This issue proposes such a template and converts each
+operand to that template.
+
+DWARF operations potentially have three sources of input, operands
+which are encoded in their byte stream, values on the stack, and the
+context in which they are executed. They also can also modify the
+stack and leave items on the stack. For every operand each of these
+sources of input, output, and side effecs are specified in a
+standardized format similar to how other stack machine operations have
+been documented in the past.
+
+Proposed Changes
+----------------
+
+In Chapter 3 replace the following paragraph:
+
+> A DWARF expression is encoded as a stream of operations, each
+> consisting of an opcode followed by zero or more literal operands. The
+> number of operands is implied by the opcode.
+
+With:
+
+> A DWARF expression is encoded as a stream of operations, each
+> operation consisting of an opcode followed by zero or more literal
+> operands. The number of operands is implied by the opcode. An may
+> also recieve parameters from the stack and make use of information
+> from its evaluation context.
+
+> *The description of the inputs of each operation will be as followes:
+> DW_OP_name
+>   operands:
+>     first_operand [encoding type]
+>     second_operand [encodeing type]
+>   stack parameters:
+>     T [expected types]: <optional> parameter1_name   # 4th entry on the stack
+>     Z [expected types]: <optional> parameter2_name   # 3rd entry on the stack
+>     Y [expected types]: <optional>  parameter3_name  # 2nd entry on stack
+>     X [expected types]: <optional> parameter4_name   # top of stack
+>   stack output:
+>     X [resulting type]
+>
+
+> The top four elements of the stack are given the names X, Y, Z, T as
+> a matter of desciptive typographical convienence. In these
+> descriptions the stack is presented as growing down with the deepest
+> entry presented first. Unless otherwise stated, it is assumed that
+> all stack parameters are consumed by the operation amd all stack
+> output is pushed onto the stack.*
+
+In Section 3.2 replace the operation descriptions as follows:
+
+> 1. `DW_OP_dup
+>   stack parameters:
+>     X [*any*]
+>   stack output:
+>     X: X
+>     Y: X
+> 
+>    The DW_OP_dup operation duplicates the entry at the top of the stack.
```

Comment:
I really thought about adding an explict "the stack parameter is not popped" to these items since I said that it was assumed above but in the end I decided that "the people reading this are computer scientists, they are smart and will understand that dup and the other explicit stack operations don't pop their paramters.